google_reauth/reauth.py

```diff
@@ -0,0 +1,296 @@
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A module that provides functions for handling rapt authentication.
+
+Reauth is a process of obtaining additional authentication (such as password,
+security token, etc.) while refreshing OAuth 2.0 credentials for a user.
+
+Credentials that use the Reauth flow must have the reauth scope,
+``https://www.googleapis.com/auth/accounts.reauth``.
+
+This module provides a high-level function for executing the Reauth process,
+:func:`refresh_access_token`, and lower-level helpers for doing the individual
+steps of the reauth process.
+
+Those steps are:
+
+1. Obtaining a list of challenges from the reauth server.
+2. Running through each challenge and sending the result back to the reauth
+   server.
+3. Refreshing the access token using the returned rapt token.
+"""
+
+import json
+import sys
+
+from six.moves import http_client
+from google_reauth import challenges
+from google_reauth import errors
+from google_reauth import _helpers
+from google_reauth import _reauth_client
+
+
+REAUTH_SCOPE = 'https://www.googleapis.com/auth/accounts.reauth'
+
+REAUTH_NEEDED_ERROR = 'invalid_grant'
+REAUTH_NEEDED_ERROR_INVALID_RAPT = 'invalid_rapt'
+REAUTH_NEEDED_ERROR_RAPT_REQUIRED = 'rapt_required'
+
+
+_challenges = challenges.build_challenges()
+
+
+def _run_next_challenge(msg, http_request, access_token):
+    """Get the next challenge from msg and run it.
+
+    Args:
+        msg: Reauth API response body (either from the initial request to
+            https://reauth.googleapis.com/v2/sessions:start or from sending the
+            previous challenge response to
+            https://reauth.googleapis.com/v2/sessions/id:continue)
+        http_request: callable to run http requests. Accepts uri, method, body
+            and headers. Returns a tuple: (response, content)
+        access_token: reauth access token
+
+    Returns: rapt token.
+    Raises:
+        errors.ReauthError if reauth failed
+    """
+    for challenge in msg['challenges']:
+        if challenge['status'] != 'READY':
+            # Skip non-activated challneges.
+            continue
+        c = _challenges[challenge['challengeType']]
```

Comment:
This should be robust in the face of new challenge types added on the server side, suggest:

```
c = _challenges.get(challenge['challengeType'], None)

if not c:
   continue
```